### PR TITLE
WAZO-3767-SAML-ACS should always redirect 

### PR DESCRIPTION
**Pull request by jirkah**

### CODE REVIEW

**Comment by jfbourque** [comment]
recheck

File: wazo_auth/services/saml.py [hunk 1]
```diff
@@ -117,27 +120,43 @@ def prepare_redirect_response(
...
+            relay_state,
         )
         location = [i for i in info['headers'] if i[0] == 'Location'][0][1]
         return location, saml_session_id
 
     def process_auth_response(
         self, url: str, remote_addr: str, form_data: SAMLACSFormData
     ) -> str | None:
-        saml_client = self.get_client(form_data['RelayState'])
+        req_id: list[Any] = [
```
Comment (DrPyser) [inline wazo_auth/services/saml.py]: `list[SamlAuthContext]`

File: wazo_auth/services/saml.py [hunk 2]
```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
...
+    def _find_session_by_relay_state(
+        self, relay_state: str
+    ) -> tuple[SamlAuthContext, SamlAuthContext] | tuple[None, None]:
+        session: list[SamlAuthContext] = [
+            e
+            for e in self._outstanding_requests
+            if self._outstanding_requests[e].relay_state == relay_state
+        ]
+        if len(session) == 1:
+            return self._outstanding_requests[session[0]], session[0]
```
Comment (DrPyser) [inline wazo_auth/services/saml.py:161]:
```suggestion
        (reqid, context), *_: list[tuple[Any, SamlAuthContext]] = [
            reqid, context
            for reqid, context in self._outstanding_requests.items()
            if context.relay_state == relay_state
        ]
        if not _:
            return context, reqid
```

alternative formulation using sequence expansion assignments, avoiding sequence indexing and allowing for more intuitive variable juggling(IMHO).

File: wazo_auth/services/saml.py [hunk 3]
```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
...
+        else:
+            logger.warning(
+                "Unable to get SAML session corresponding to the received RelayState"
+            )
+            return None, None
+
+    def _process_auth_response_error(
+        self, redirect_url: str, req_id: Any, msg: str
```
Comment (DrPyser) [inline wazo_auth/services/saml.py:169]: typing suggestion: define a type alias `RequestId = Any` and use `RequestId` in annotations. 
This helps make the annotations more readable, and we can change the type alias in a single place if we discover a more appropriate typing.

```suggestion
        self, redirect_url: str, req_id: RequestId, msg: str
```

File: wazo_auth/services/saml.py [hunk 4]
```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
...
+            saml_response,
+            BINDING_HTTP_POST,
+            self._outstanding_requests,
+            None,
+            conv_info=conv_info,
+        )
+
+    def _find_session_by_relay_state(
+        self, relay_state: str
+    ) -> tuple[SamlAuthContext, SamlAuthContext] | tuple[None, None]:
```
Comment (DrPyser) [inline wazo_auth/services/saml.py:154]:
```suggestion
    ) -> tuple[SamlAuthContext, Requestid] | tuple[None, None]:
```
error in type annotation

**Comment by DrPyser** [comment]
During manual tests with `failed.com` domain configuration:
```
2024-06-21 18:22:08,657 [1889673] (INFO) (wazo-auth): request: POST https://stack.wazo.local/api/auth/0.1/saml/sso {'Host': 'stack.wazo.local', 'X-Script-Name': '/api/auth', 'X-Forwarded-For': '10.0.4.6', 'X-Forwarded-Proto': 'https', 'Connection': 'close', 'Sec-Ch-Ua': '"Not/A)Brand";v="8", "Chromium";v="126"', 'Sec-Ch-Ua-Platform': '"Linux"', 'Sec-Ch-Ua-Mobile': '?0', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.0.0 Safari/537.36', 'Accept': '*/*', 'Origin': 'https://app.wazo.local', 'Sec-Fetch-Site': 'same-site', 'Sec-Fetch-Mode': 'cors', 'Sec-Fetch-Dest': 'empty', 'Referer': 'https://app.wazo.local/', 'Accept-Encoding': 'gzip, deflate, br, zstd', 'Accept-Language': 'en-CA,en;q=0.9,fr-CA;q=0.8,fr;q=0.7,en-GB;q=0.6', 'Content-Type': 'application/json', 'Content-Length': '76'} with data {"redirect_url": "https://app.wazo.local/postacs.html", "domain": "failed.com"}                                                                                                                                     2024-06-21 18:22:08,661 [1889673] (ERROR) (wazo_auth.plugins.http.saml.http): Failed to process initial SAML SSO post because of: 'failed.com'
2024-06-21 18:22:08,662 [1889673] (ERROR) (xivo.rest_api_helpers): SAML client for domain not found or failed: {'domain': 'failed.com'}
2024-06-21 18:22:08,664 [1889673] (INFO) (wazo-auth): response to 10.0.4.6 in 0.00s: POST https://stack.wazo.local/api/auth/0.1/saml/sso 500
2024-06-21 18:22:08,666 [1889673] (DEBUG) (wazo-auth): response body: {"message": "SAML client for domain not found or failed", "error_id": "configuration-error", "details": {"domain": "failed.com"}, "timestamp": 1718994128.6626027, "resource": "saml"}
```
This does not result in a redirect.

**Comment by DrPyser** [comment]
Otherwise, could verify that with a properly configured tenant associated with the domain, the failure of the SAML response validation still leads to a redirect to `https://app.wazo.local/postacs.html?login_failure_code=500`

File: wazo_auth/services/saml.py [hunk 5]
```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
...
+        logger.warning(msg)
+        logger.debug(f'Removing session: {req_id}')
+        del self._outstanding_requests[req_id]
+        raise exceptions.SAMLProcessingErrorWithReturnURL(
+            'Unknown principal', return_url=redirect_url
+        )
+
+    def _process_auth_response_context_not_found(self) -> None:
+        logger.warning('ACS response request failed: Context not found')
+        raise exceptions.SAMLProcessingError('Context not found', code=404)
```
Comment (DrPyser) [inline wazo_auth/services/saml.py:181]: Not sure I see the value of encapsulating a log and an exception. They would benefit from being embedded in the context where they should occur. 
Hiding exceptions behind method calls leads to code being harder to read(you don't know by seeing a method call that this is actually unconditionally raising an exception).

File: wazo_auth/services/saml.py [hunk 6]
```diff
@@ -117,57 +123,148 @@ def prepare_redirect_response(
...
+    ) -> None:
+
+        logger.warning(msg)
+        logger.debug(f'Removing session: {req_id}')
+        del self._outstanding_requests[req_id]
+        raise exceptions.SAMLProcessingErrorWithReturnURL(
+            'Unknown principal', return_url=redirect_url
+        )
+
+    def _process_auth_response_context_not_found(self) -> None:
```
Comment (DrPyser) [inline wazo_auth/services/saml.py:179]:
```suggestion
    def _process_auth_response_context_not_found(self) -> NoReturn:
```
functions that never returns(e.g. always raise) should use `NoReturn` annotation to tell the type checker.

**Review by DrPyser — APPROVED** [review]
Well done! Let's merge

**Pull request merged**